### ISSUE TITLE
`parse` mutates object and adds a `times` property to it

### ISSUE THREAD

**Issue by wmartins** [colinhacks/zod#433]
Hey there! I think I've got an edge case on using a combination of Zod's parse with recursive schema definition.

Basically, I have a schema that uses itself recursively in order to match against a JSON object. It parses correctly, however, I noticed that after parsing, it mutates the original object, adding a property `times` with the value as `NaN`.

Here's the example:

```ts
import assert from "assert";
import * as z from "zod";

export type None = undefined | null;
export namespace None {
  export const schema: z.ZodSchema<None> = z.union([z.undefined(), z.null()]);
}

export type Some = string | number | boolean;
export namespace Some {
  export const schema: z.ZodSchema<Some> = z.union([z.string(), z.number(), z.boolean()]);
}

export type Value = Some | Data | None;
export namespace Value {
  export const schema: z.ZodSchema<Value> = z.lazy(() => z.union([Some.schema, Data.schema, None.schema]));
}

export type Data = Record<string, unknown>;
export namespace Data {
  export const schema: z.ZodSchema<Data> = z.record(Value.schema);
}

const clone = (data: Object) => JSON.parse(JSON.stringify(data));

const data = {
  payload: {
    data: {},
  },
};

const original = clone(Object.freeze(data));

const result = Data.schema.parse(data);

assert.deepStrictEqual(original, result);
```

In this case the assertion fails due to `parse` have mutated this object, adding `times` to it:

```
  {
    properties: {
      data: {},
-     times: NaN
    }
  }
```

I took a look to try to understand why and it seems the culprit (or at least it's one step into fully understanding why) is this little part here:

https://github.com/colinhacks/zod/blob/1c51624d32756804a0b02a2e18eab31f9234ec1f/src/parser.ts#L133-L152

Basically, as my property is also named `data`, I think we must have some namespaces/variables conflict. When I test with changing `data` to `myData`, it works just fine.

**Comment by wmartins** [comment]
Took a stab at writing a test case for:

```ts
test('does not mutates nested schema objects', () => {
  const primitives = z.union([z.string(), z.number(), z.boolean()]);

  type Entries = Record<string, unknown>;
  const entries: z.ZodSchema<Entries> = z.lazy(() =>
    z.record(z.union([primitives, entries])),
  );

  const parsed = entries.parse({
    nested: {
      data: {},
    },
  });

  expect(parsed).toEqual({ nested: { data: {} } });
});
```

And I also confirmed that namespacing _data_ probably makes it work:

```diff
diff --git a/src/parser.ts b/src/parser.ts
index 3496589..3358899 100644
--- a/src/parser.ts
+++ b/src/parser.ts
@@ -9,10 +9,12 @@ import {
 import { util } from './helpers/util';
 import { ZodErrorMap, defaultErrorMap } from './defaultErrorMap';
 
+const namespacedData = Symbol('data');
+
 export type ParseParams = {
   seen?: {
     schema: any;
-    objects: { data: any; error?: any; times: number }[];
+    objects: { [namespacedData]: any; error?: any; times: number }[];
   }[];
   path?: (string | number)[];
   errorMap?: ZodErrorMap;
@@ -132,7 +134,7 @@ export const ZodParser = (schemaDef: z.ZodTypeDef) => (
 
   if (isNonprimitive) {
     if (schemaSeen) {
-      const found = find(schemaSeen.objects, x => x.data === obj);
+      const found = find(schemaSeen.objects, x => x[namespacedData] === obj);
 
       if (found) {
         if (found.error) {
@@ -152,7 +154,7 @@ export const ZodParser = (schemaDef: z.ZodTypeDef) => (
     } else {
       params.seen.push({
         schema: schemaDef,
-        objects: [{ data: obj, error: undefined, times: 1 }],
+        objects: [{ [namespacedData]: obj, error: undefined, times: 1 }],
       });
     }
   }
```

In case that makes any sense, I can open a PR to fix that (just need to figure out a better name for that `namespacedData` lol).

**Comment by colinhacks** [comment]
Good call. My recommended fix is to switch to Zod 3. It's about to replace v1 so issues with v1 are very low priority. Should be an easy switch, there are very few breaking changes.

**Comment by wmartins** [comment]
@colinhacks, thank you for your answer! Just saw the upgrade guide and I'll take a look into how to upgrade our codebase.

In the meantime, is there a deprecation plan for zod v1? Just because we use this version in a couple of places
